File: 0701/SeoyeonHong/메뉴 리뉴얼.py
# https://school.programmers.co.kr/learn/courses/30/lessons/72411
from collections import defaultdict
from itertools import combinations

def solution(orders, course):
    answer = []
    count = [defaultdict(int) for _ in range(len(course))]
    
    for order in orders:
        order = sorted(order)
        for c in range(len(course)):
            if course[c] <= len(order):
                for combination in combinations(order, course[c]):
                    count[c][''.join(combination)] += 1
    
    options = [list(count[i].items()) for i in range(len(course))]

    for option in options:
        if option:
            option.sort(key=lambda x: (-x[1], x[0]))
            if option[0][1] >= 2:
                for candidate, ordered in option:
                    if ordered == option[0][1]:
                        answer.append(candidate)
                    else:
                        break
    return sorted(answer)

File: 0701/SoohyunKim/메뉴리뉴얼.js
// https://school.programmers.co.kr/learn/courses/30/lessons/72411

function solution(orders, course) {
    const answer = []
    
    for (const cnt of course) {
        const map = new Map()
        for (const order of orders) {
            getCombinations(order.split("").sort(), cnt).forEach(el => {
                // 메뉴 조합 횟수 기록
                el = el.join("")
                if (map.get(el)) map.set(el , map.get(el)+1)
                else map.set(el, 1)
            })
        }
        
        let tmpArr = []
        let max = 2
        for (const [key, val] of map) {
            if (val > max) {
                max = val
                tmpArr = [key]
            } else if (val === max) {
                tmpArr.push(key)
            }
        }
        answer.push(...tmpArr)
    }
    
    return answer.sort()
}

// 조함 생성
function getCombinations(arr, n) {
    let res = []
    if(n===1) return arr.map(el =>[el]);
    
    arr.forEach((fixed, idx, arr) => {
        const rest = arr.slice(idx+1)
        const combination = getCombinations(rest, n-1)    
        const attached = combination.map(el => [fixed, ...el])
        res.push(...attached)
    })
    
    return res;
}